Add simple locking to the internal Tweak data

ptschip opened the pull request.

ptschip committed:
Add simple locking for tweak object internal data

ptschip committed:
Make internal Tweak data private and add accessor methods

awemany commented on src/tweak.h:
```diff
@@ -117,23 +123,36 @@ class CTweakRef : public CTweakBase
+        std::unique_lock<std::mutex> lck(cs_tweak);
+        return name;
+    }
+    virtual std::string GetHelp() const
```
What's the reason to lock getting the `help` and `name` of the tweak? Isn't that ialways expected to be immutable data that can accessed in parallel?

(Not that this would likely be any performance issue either way, just wondering)

ptschip committed:
Remove lock from accessing help or name and make them const

ptschip commented:
thanks, took of the lock and made them const

On 03/08/2018 3:51 AM, awemany wrote:
>
> *@awemany* commented on this pull request.
>
> ------------------------------------------------------------------------
>
> In src/tweak.h
> <https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/1237#discussion_r207509511>:
>
> >          : name(namep), help(helpp), value(val), eventCb(callback)
>      {
>          tweaks[CTweakKey(name)] = this;
>      }
>  
> -    virtual std::string GetName() const { return name; }
> -    virtual std::string GetHelp() const { return help; }
> -    virtual UniValue Get() const { return UniValue(*value); }
> +    virtual std::string GetName() const
> +    {
> +        std::unique_lock<std::mutex> lck(cs_tweak);
> +        return name;
> +    }
> +    virtual std::string GetHelp() const
>
> What's the reason to lock getting the |help| and |name| of the tweak?
> Isn't that ialways expected to be immutable data that can accessed in
> parallel?
>
> (Not that this would likely be any performance issue either way, just
> wondering)
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/1237#pullrequestreview-143136899>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AMRF0E5NPSQnfk6tNz0xJKp36bJZgU6cks5uNCs0gaJpZM4VtDnp>.
>

awemany commented on src/tweak.h:
```diff
@@ -11,6 +11,10 @@
 // c++11 #include <type_traits>
 #include "univalue/include/univalue.h"
 
+#include <iostream>
```
Nit: Do you need `<iostream>` or is that just a relic from some debugging?

awemany reviewed (approved):
Looks good to me. What is the intent of this change? Just general code cleanup?

ptschip committed:
Remove unnecessary include

ptschip committed:
Switch to slightly more efficient std::lock_guard

We have no need to unlock and relock at any point so there is no
need for using std:unique_lock here.
